### DISCUSSION TITLE
Docs: Direction for using `@next/font` with React Portal / Headless UI Dialogs

### DISCUSSION THREAD

**Discussion by frankfka** [vercel/next.js#70724] [Help]
### What is the improvement or update you wish to see?

Indicate the recommended approach to use `@next/font` with Headless UI Dialogs / other applications of React Portal that inject an element directly as a child of `body`.

### Is there any context that might help us understand?

The `@next/font` documentation suggests that the `className` (or `variable`) on the `main` component within `_app`. However, this means that the font is not accessible within direct children of `body`, as `body` is rendered within `_document`.

This becomes an issue with Headless UI dialogs, which injects the dialog directly as a child of `body`. Here's the [source](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/components/portal/portal.tsx#L26).

We found two workarounds for this:
1. Manually specify the font `className`/`variable` directly in the root `Dialog` component
2. Instantiate the font in `_app`, export it as a `const`, then import it in `_document` and add the relevant `className` to `body`. For reasons unknown to me, this only works if the font is exported in `_app`, and not in a util file.

### Does the docs page already exist? Please link to it.

https://nextjs.org/docs/basic-features/font-optimization

**Comment by emin-alizada** [comment]
I have encountered the same problem

**Comment by hanneslund** [comment]
You could export your font in a seperate file and import it in `_app` and your `Dialog` component.
```jsx
// fonts.ts
import { Raleway } from '@next/font/google'

export const font = Raleway({  subsets: ['latin'] })
```

Or you could set it on `html` like [this](https://nextjs.org/docs/basic-features/font-optimization#apply-the-font-in-head).

**Comment by stefanprobst** [comment]
any reason why `next/font` does not simply set the css variable on `:root` by default? i.e.:

```ts
Inter({ variable: "--font-inter" });
```

would inject:

```css
:root { 
  --font-inter: '__Inter_8224f7', '__Inter_Fallback_8224f7';
}
```

so we could do:

```css
/* src/styles/globals.css */
html {
  font-family: var(--font-inter), system-ui, sans-serif;
}
```

**Comment by derekaug** [comment]
> any reason why `next/font` does not simply set the css variable on `:root` by default? i.e.:
> 
> ```ts
> Inter({ variable: "--font-inter" });
> ```
> 
> would inject:
> 
> ```css
> :root { 
>   --font-inter: '__Inter_8224f7', '__Inter_Fallback_8224f7';
> }
> ```
> 
> so we could do:
> 
> ```css
> /* src/styles/globals.css */
> html {
>   font-family: var(--font-inter), system-ui, sans-serif;
> }
> ```

I'm curious as to this as well, I ended up adding what is essentially equivalent to the below snippet in my `_app.tsx` to work around this and be able to use the variable in my `global.css` (and use it with tailwind as the default font).

```tsx
<style jsx global>{`
  :root {
    --font-inter: ${Inter.style.fontFamily};
  }
`}</style>
```

**Comment by dennis-gonzales** [comment]
> > any reason why `next/font` does not simply set the css variable on `:root` by default? i.e.:
> > ```ts
> > Inter({ variable: "--font-inter" });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > would inject:
> > ```css
> > :root { 
> >   --font-inter: '__Inter_8224f7', '__Inter_Fallback_8224f7';
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so we could do:
> > ```css
> > /* src/styles/globals.css */
> > html {
> >   font-family: var(--font-inter), system-ui, sans-serif;
> > }
> > ```
> 
> I'm curious as to this as well, I ended up adding what is essentially equivalent to the below snippet in my `_app.tsx` to work around this and be able to use the variable in my `global.css` (and use it with tailwind as the default font).
> 
> ```tsx
> <style jsx global>{`
>   :root {
>     --font-inter: ${Inter.style.fontFamily};
>   }
> `}</style>
> ```

This works for me splendidly, thanks.

**Comment by blakeley** [comment]
Unfortunately, this doesn't seem to work with server components, resulting in the following error:

```
- error ./app/layout.tsx
'client-only' cannot be imported from a Server Component module. It should only be used from a Client Component.

The error was caused by using 'styled-jsx' in './app/layout.tsx'. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.

Import trace for requested module:
  ./app/layout.tsx
```

Ideally, we need a solution that works in server components, too.

**Comment by joshbedo** [comment]
was there ever a resolution for this? having the same issue

**Comment by ludwighogstrom** [comment]
Same problem here! 

What's the technical reason why the font can't be added to body in `_document`? Feels like the most natural way.

**Comment by samcx** [comment]
Hi everyone—

Just moving this to [Discussions](https://github.com/vercel/next.js/discussions) as the original ask was a Documentation clarification request.
